[WIP] Create a getCallNodeSamplesByDepth function for flame graphs

brisad opened the pull request:
WIP for #604 

Create `CallNodeSamplesByDepth`, based on `StackTimingByDepth` for the stack chart.
This is a proposal of how a structure representing a flame graph could look like. Please let me know if this is a viable solution.

Instead of letting the boxes in a flame graph represent different stacks, I thought it would be more appropriate to let them represent different call nodes. Otherwise we'd see the same function represented as disjoint boxes in the graph if we have samples from the same function but from different stacks. Using call nodes, those boxes will get merged, which I think makes more sense for a flame graph. And it makes it more readable.

brisad commented on src/profile-logic/call-node-samples.js:
```diff
@@ -0,0 +1,221 @@
+
+export type CallNodeSamplesByDepth = Array<{
+  start: Milliseconds[],
+  end: Milliseconds[],
```
The x-axis of a flame graph does not represent time. So these units should probably change, and can perhaps instead be represented as some kind of width. But I don't want to make too many changes before I know that I am on the right track.

brisad commented on src/profile-logic/call-node-samples.js:
```diff
@@ -0,0 +1,221 @@
+      sampleTime
+    );
+    previousDepth = depth;
+    sampleTime += samples;
```
We're just counting samples here, so the name `sampleTime` should probably change.

brisad commented on src/test/store/actions.test.js:
```diff
@@ -140,6 +140,25 @@ describe('selectors/getStackTimingByDepthForStackChart', function() {
   });
 });
 
+describe('selectors/getCallNodeSamplesByDepthForFlameGraph', function() {
```
There should of course be a comment here with some nice illustration showing what the flame graph should look like.

gregtatum commented:
Nice!

I rewrote the test a bit so I can verify what was going on. I think everything is computing properly, except an expectation of the ordering of the samples. I think @mstange would prefer the heavier weighted samples to be to the left of the graph. My last test sample has that assertion, and it's failing on the test.

The call tree already computes the structure

```diff
diff --git a/src/test/store/actions.test.js b/src/test/store/actions.test.js
index c68e17f..e5263a3 100644
--- a/src/test/store/actions.test.js
+++ b/src/test/store/actions.test.js
@@ -17,6 +17,7 @@ import {
 } from '../../actions/profile-view';
 import { changeStackChartColorStrategy } from '../../actions/stack-chart';
 import { getCategoryByImplementation } from '../../profile-logic/color-categories';
+import { getProfileFromTextSamples } from '../fixtures/profiles/make-profile';
 
 const { selectedThreadSelectors } = ProfileViewSelectors;
 
@@ -141,20 +142,85 @@ describe('selectors/getStackTimingByDepthForStackChart', function() {
 });
 
 describe('selectors/getCallNodeSamplesByDepthForFlameGraph', function() {
-  it('computes call node samples by depth', function() {
-    const store = storeWithProfile();
+  /**
+   * Map the callNodeSamplesByDepth data structure into a human readable format where
+   * each line takes the form:
+   *
+   * "FunctionName1 (StartTime:EndTime) | FunctionName2 (StartTime:EndTime)"
+   */
+  function getHumanReadableCallNodeSamplesByDepth(store, funcNames) {
+    const { callNodeTable } = selectedThreadSelectors.getCallNodeInfo(
+      store.getState()
+    );
     const callNodeSamplesByDepth = selectedThreadSelectors.getCallNodeSamplesByDepthForFlameGraph(
       store.getState()
     );
 
-    expect(callNodeSamplesByDepth).toEqual([
-      { start: [0], end: [10], callNode: [0], length: 1 },
-      { start: [0], end: [9], callNode: [1], length: 1 },
-      { start: [0, 2, 3], end: [2, 3, 7], callNode: [2, 3, 4], length: 3 },
-      { start: [3], end: [5], callNode: [5], length: 1 },
-      { start: [3], end: [4], callNode: [6], length: 1 },
-      { start: [3], end: [4], callNode: [7], length: 1 },
-      { start: [3], end: [4], callNode: [8], length: 1 },
+    return callNodeSamplesByDepth.map(({ callNode, end, length, start }) => {
+      const lines = [];
+      for (let i = 0; i < length; i++) {
+        const callNodeIndex = callNode[i];
+        const funcIndex = callNodeTable.func[callNodeIndex];
+        const funcName = funcNames[funcIndex];
+        lines.push(`${funcName} (${start[i]}:${end[i]})`);
+      }
+      return lines.join(' | ');
+    });
+  }
+
+  it('computes a basic example', function() {
+    const { profile, funcNames } = getProfileFromTextSamples(`
+      A A A
+      B B B
+      C C H
+      D F I
+      E G
+    `);
+
+    const store = storeWithProfile(profile);
+    expect(getHumanReadableCallNodeSamplesByDepth(store, funcNames)).toEqual([
+      'A (0:3)',
+      'B (0:3)',
+      'C (0:2) | H (2:3)',
+      'D (0:1) | F (1:2) | I (2:3)',
+      'E (0:1) | G (1:2)',
+    ]);
+  });
+
+  it('can handle null samples', function() {
+    const { profile, funcNames } = getProfileFromTextSamples(`
+      A A X A
+      B B   B
+      C C   H
+      D F   I
+      E G
+    `);
+
+    // Remove the X sample by setting it's stack to null.
+    profile.threads[0].samples.stack[2] = null;
+
+    const store = storeWithProfile(profile);
+    expect(getHumanReadableCallNodeSamplesByDepth(store, funcNames)).toEqual([
+      'A (0:3)',
+      'B (0:3)',
+      'C (0:2) | H (2:3)',
+      'D (0:1) | F (1:2) | I (2:3)',
+      'E (0:1) | G (1:2)',
+    ]);
+  });
+
+  it('weights heavier samples to the left', function() {
+    const { profile, funcNames } = getProfileFromTextSamples(`
+      A D D D
+      B E F F
+      C     G
+    `);
+
+    const store = storeWithProfile(profile);
+    expect(getHumanReadableCallNodeSamplesByDepth(store, funcNames)).toEqual([
+      'D (0:3) | A (3:4)',
+      'F (0:2) | E (2:3) | B (3:4)',
+      'G (0:1) | C (3:4)',
     ]);
   });
 });

```

I probably wasn't explicit about that failing case, but I think we can leverage the CallTree to simplify the creation of the timing data structure. I was thinking of something like the following. I didn't run this code, so I'm not sure if it actually works. It would be better as a loop rather than a recursive function, but I wanted to slam out what I was thinking:

```js
    const timing = [];
    function computeDepth(nodeIndexes, timeOffsetParent = 0, depth = 0) {
      if (nodeIndexes.length === 0) {
        return;
      }

      // Select an existing row, or create a new one.
      let row = timing[depth];
      if (row === undefined) {
        row = {
          start: [],
          end: [],
          callNode: [],
          length: 0,
        };
        timing[depth] = row;
      }

      // Compute the timing information.
      let timeOffset = timeOffsetParent;
      for (let i = 0; i < nodeIndexes.length; i++) {
        const nodeIndex = nodeIndexes[i];
        const node = callTree.getNode(nodeIndex);
        row.start.push(timeOffset);
        row.end.push(timeOffset + node.totalTime);
        row.callNode.push(nodeIndex);
        row.length++;
        timeOffset += node.totalTime;
        computeDepth(callTree.getChildren(nodeIndex), timeOffset, depth + 1);
      }
    }
    computeDepth(callTree.getRoots());
```

I think the naming of the data structure could be simplified from `CallNodeSamplesByDepth` to `FlameGraphTiming`.

Eventually, the timing information may be nice to have as a unit interval (ranged from 0 to 1) rather than milliseconds. The call tree already computes this with the "percentage" field, but it's plaintext, and not a float number. We could potentially tweak that and get the unit interval feature for almost free.

All that said, awesome work so far!

mstange commented:
> I think @mstange would prefer the heavier weighted samples to be to the left of the graph.

I haven't read the patch, but [the canonical "Flame Graph"](http://www.brendangregg.com/flamegraphs.html) sorts stacks alphabetically by function name. We should do that.

Sorting by function name means that the sorting is consistent even as the workload and the shape of the profile changes. This can make it easier to compare flame graphs visually.

gregtatum commented:
Thanks for the clarification! That's a really good point.

brisad commented:
Thanks for the great comments! Nice with that human readable part of the tests. And wow, leveraging the `CallTree` really simplified things!

I've made an iterative version of your example code and renamed the structure as suggested. I also updated your last test to work with alphabetical ordering. Will upload in a few minutes. Please let me know what you think.

Will look into the unit interval part next, but wanted to share this part first.

And sorry for the long delay :-(

gregtatum commented:
No worries on the delay! I've been out on holidays for the past week so I'm just now back at my computer.

brisad commented on src/profile-logic/flame-graph.js:
```diff
@@ -0,0 +1,77 @@
+
+  while (stack.length) {
+    const { depth, nodeIndex } = stack.pop();
+    const totalTime = callTree._callNodeTimes.totalTime[nodeIndex];
```
`getNode` returns data formatted as strings, so for now I just reached into this table directly but it should of course be dealt with in a better way. Can we extend the object returned from `getNode` to include numeric data, or perhaps split the method into two, one with formatted data and one without?

gregtatum replied:
Feel free to modify `getNode` to include the data you need, or wherever feels logical to you.

brisad commented:
Created a new (non WIP-tagged) PR #672 to replace this one.
